// botframework-connector/lib/tokenApi/operations/botSignIn.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/botSignInMappers");
const Parameters = require("../models/parameters");
/** Class representing a BotSignIn. */
class BotSignIn {
    /**
     * Create a BotSignIn.
     * @param {TokenApiClientContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    getSignInUrl(state, options, callback) {
        return this.client.sendOperationRequest({
            state,
            options
        }, getSignInUrlOperationSpec, callback);
    }
}
exports.BotSignIn = BotSignIn;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const getSignInUrlOperationSpec = {
    httpMethod: "GET",
    path: "api/botsignin/GetSignInUrl",
    queryParameters: [
        Parameters.state,
        Parameters.codeChallenge,
        Parameters.emulatorUrl,
        Parameters.finalRedirect
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "String"
                }
            }
        },
        default: {}
    },
    serializer
};
