// botframework-connector/lib/tokenApi/operations/userToken.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const msRest = require("@azure/ms-rest-js");
const Mappers = require("../models/userTokenMappers");
const Parameters = require("../models/parameters");
/** Class representing a UserToken. */
class UserToken {
    /**
     * Create a UserToken.
     * @param {TokenApiClientContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    getToken(userId, connectionName, options, callback) {
        return this.client.sendOperationRequest({
            userId,
            connectionName,
            options
        }, getTokenOperationSpec, callback);
    }
    getAadTokens(userId, connectionName, aadResourceUrls, options, callback) {
        return this.client.sendOperationRequest({
            userId,
            connectionName,
            aadResourceUrls,
            options
        }, getAadTokensOperationSpec, callback);
    }
    signOut(userId, options, callback) {
        return this.client.sendOperationRequest({
            userId,
            options
        }, signOutOperationSpec, callback);
    }
    getTokenStatus(userId, options, callback) {
        return this.client.sendOperationRequest({
            userId,
            options
        }, getTokenStatusOperationSpec, callback);
    }
}
exports.UserToken = UserToken;
// Operation Specifications
const serializer = new msRest.Serializer(Mappers);
const getTokenOperationSpec = {
    httpMethod: "GET",
    path: "api/usertoken/GetToken",
    queryParameters: [
        Parameters.userId,
        Parameters.connectionName0,
        Parameters.channelId,
        Parameters.code
    ],
    responses: {
        200: {
            bodyMapper: Mappers.TokenResponse
        },
        404: {
            bodyMapper: Mappers.TokenResponse
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const getAadTokensOperationSpec = {
    httpMethod: "POST",
    path: "api/usertoken/GetAadTokens",
    queryParameters: [
        Parameters.userId,
        Parameters.connectionName0,
        Parameters.channelId
    ],
    requestBody: {
        parameterPath: "aadResourceUrls",
        mapper: Object.assign({}, Mappers.AadResourceUrls, { required: true })
    },
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "Composite",
                            className: "TokenResponse"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const signOutOperationSpec = {
    httpMethod: "DELETE",
    path: "api/usertoken/SignOut",
    queryParameters: [
        Parameters.userId,
        Parameters.connectionName1,
        Parameters.channelId
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Object"
                }
            }
        },
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
const getTokenStatusOperationSpec = {
    httpMethod: "GET",
    path: "api/usertoken/GetTokenStatus",
    queryParameters: [
        Parameters.userId,
        Parameters.channelId,
        Parameters.include
    ],
    responses: {
        200: {
            bodyMapper: {
                serializedName: "parsedResponse",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "TokenStatus"
                        }
                    }
                }
            }
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer
};
