// botbuilder-dialogs/lib/componentDialog.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module botbuilder-dialogs
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
const botbuilder_core_1 = require("botbuilder-core");
const dialog_1 = require("./dialog");
const dialogContext_1 = require("./dialogContext");
const dialogContainer_1 = require("./dialogContainer");
const PERSISTED_DIALOG_STATE = 'dialogs';
/**
 * Base class for a dialog that contains other child dialogs.
 *
 * @remarks
 * Component dialogs let you break your bot's logic up into components that can themselves be added
 * as a dialog to another `ComponentDialog` or `DialogSet`. Components can also be exported as part
 * of a node package and used within other bots.
 *
 * To define a new component derive a class from ComponentDialog and add your child dialogs within
 * the classes constructor:
 *
 * ```JavaScript
 * const { ComponentDialog, WaterfallDialog, TextPrompt, NumberPrompt } = require('botbuilder-dialogs');
 *
 * class FillProfileDialog extends ComponentDialog {
 *     constructor(dialogId) {
 *         super(dialogId);
 *
 *         // Add control flow dialogs
 *         this.addDialog(new WaterfallDialog('start', [
 *             async (step) => {
 *                 // Ask user their name
 *                 return await step.prompt('namePrompt', `What's your name?`);
 *             },
 *             async (step) => {
 *                 // Remember the users answer
 *                 step.values['name'] = step.result;
 *
 *                 // Ask user their age.
 *                 return await step.prompt('agePrompt', `Hi ${step.values['name']}. How old are you?`);
 *             },
 *             async (step) => {
 *                 // Remember the users answer
 *                 step.values['age'] = step.result;
 *
 *                 // End the component and return the completed profile.
 *                 return await step.endDialog(step.values);
 *             }
 *         ]));
 *
 *         // Add prompts
 *         this.addDialog(new TextPrompt('namePrompt'));
 *         this.addDialog(new NumberPrompt('agePrompt'))
 *     }
 * }
 * module.exports.FillProfileDialog = FillProfileDialog;
 * ```
 *
 * You can then add new instances of your component to another `DialogSet` or `ComponentDialog`:
 *
 * ```JavaScript
 * const dialogs = new DialogSet(dialogState);
 * dialogs.add(new FillProfileDialog('fillProfile'));
 * ```
 * @param O (Optional) options that can be passed into the `DialogContext.beginDialog()` method.
 */
class ComponentDialog extends dialogContainer_1.DialogContainer {
    beginDialog(outerDC, options) {
        return __awaiter(this, void 0, void 0, function* () {
            // Start the inner dialog.
            const innerDC = this.createChildContext(outerDC);
            const turnResult = yield this.onBeginDialog(innerDC, options);
            // Check for end of inner dialog
            if (turnResult.status !== dialog_1.DialogTurnStatus.waiting) {
                // Return result to calling dialog
                return yield this.endComponent(outerDC, turnResult.result);
            }
            // Just signal end of turn
            return dialog_1.Dialog.EndOfTurn;
        });
    }
    continueDialog(outerDC) {
        return __awaiter(this, void 0, void 0, function* () {
            // Continue execution of inner dialog.
            const innerDC = this.createChildContext(outerDC);
            const turnResult = yield this.onContinueDialog(innerDC);
            // Check for end of inner dialog
            if (turnResult.status !== dialog_1.DialogTurnStatus.waiting) {
                // Return result to calling dialog
                return yield this.endComponent(outerDC, turnResult.result);
            }
            // Just signal end of turn
            return dialog_1.Dialog.EndOfTurn;
        });
    }
    resumeDialog(dc, reason, result) {
        return __awaiter(this, void 0, void 0, function* () {
            // Containers are typically leaf nodes on the stack but the dev is free to push other dialogs
            // on top of the stack which will result in the container receiving an unexpected call to
            // resumeDialog() when the pushed on dialog ends.
            // To avoid the container prematurely ending we need to implement this method and simply
            // ask our inner dialog stack to re-prompt.
            yield this.repromptDialog(dc.context, dc.activeDialog);
            return dialog_1.Dialog.EndOfTurn;
        });
    }
    repromptDialog(context, instance) {
        return __awaiter(this, void 0, void 0, function* () {
            // Forward to inner dialogs
            const innerDC = this.createInnerDC(context, instance);
            yield innerDC.repromptDialog();
            // Notify component.
            yield this.onRepromptDialog(context, instance);
        });
    }
    endDialog(context, instance, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            // Forward cancel to inner dialogs
            if (reason === dialog_1.DialogReason.cancelCalled) {
                const innerDC = this.createInnerDC(context, instance);
                yield innerDC.cancelAllDialogs();
            }
            // Notify component
            yield this.onEndDialog(context, instance, reason);
        });
    }
    /**
     * Adds a child dialog or prompt to the components internal `DialogSet`.
     *
     * @remarks
     * The `Dialog.id` of the first child added to the component will be assigned to the [initialDialogId](#initialdialogid)
     * property.
     * @param dialog The child dialog or prompt to add.
     */
    addDialog(dialog) {
        this.dialogs.add(dialog);
        if (this.initialDialogId === undefined) {
            this.initialDialogId = dialog.id;
        }
        return this;
    }
    /**
     * Creates the inner dialog context
     * @param outerDC the outer dialog context
     */
    createChildContext(outerDC) {
        const innerDC = this.createInnerDC(outerDC.context, outerDC.activeDialog);
        innerDC.parent = outerDC;
        return innerDC;
    }
    /**
     * Called anytime an instance of the component has been started.
     *
     * @remarks
     * SHOULD be overridden by components that wish to perform custom interruption logic. The
     * default implementation calls `innerDC.beginDialog()` with the dialog assigned to
     * [initialDialogId](#initialdialogid).
     * @param innerDC Dialog context for the components internal `DialogSet`.
     * @param options (Optional) options that were passed to the component by its parent.
     */
    onBeginDialog(innerDC, options) {
        return innerDC.beginDialog(this.initialDialogId, options);
    }
    /**
     * Called anytime a multi-turn component receives additional activities.
     *
     * @remarks
     * SHOULD be overridden by components that wish to perform custom interruption logic. The
     * default implementation calls `innerDC.continueDialog()`.
     * @param innerDC Dialog context for the components internal `DialogSet`.
     */
    onContinueDialog(innerDC) {
        return innerDC.continueDialog();
    }
    /**
     * Called when the component is ending.
     *
     * @remarks
     * If the `reason` code is equal to `DialogReason.cancelCalled`, then any active child dialogs
     * will be cancelled before this method is called.
     * @param context Context for the current turn of conversation.
     * @param instance The components instance data within its parents dialog stack.
     * @param reason The reason the component is ending.
     */
    onEndDialog(context, instance, reason) {
        return Promise.resolve();
    }
    /**
     * Called when the component has been requested to re-prompt the user for input.
     *
     * @remarks
     * The active child dialog will have already been asked to reprompt before this method is called.
     * @param context Context for the current turn of conversation.
     * @param instance The instance of the current dialog.
     */
    onRepromptDialog(context, instance) {
        return Promise.resolve();
    }
    /**
     * Called when the components last active child dialog ends and the component is ending.
     *
     * @remarks
     * SHOULD be overridden by components that wish to perform custom logic before the component
     * ends.  The default implementation calls `outerDC.endDialog()` with the `result` returned
     * from the last active child dialog.
     * @param outerDC Dialog context for the parents `DialogSet`.
     * @param result Result returned by the last active child dialog. Can be a value of `undefined`.
     */
    endComponent(outerDC, result) {
        return outerDC.endDialog(result);
    }
    createInnerDC(context, instance) {
        const dialogState = instance.state[PERSISTED_DIALOG_STATE] || { dialogStack: [] };
        instance.state[PERSISTED_DIALOG_STATE] = dialogState;
        const innerDC = new dialogContext_1.DialogContext(this.dialogs, context, dialogState);
        return innerDC;
    }
    /**
     * Set the telemetry client, and also apply it to all child dialogs.
     * Future dialogs added to the component will also inherit this client.
     */
    set telemetryClient(client) {
        this._telemetryClient = client ? client : new botbuilder_core_1.NullTelemetryClient();
        this.dialogs.telemetryClient = client;
    }
    /**
     * Get the current telemetry client.
     */
    get telemetryClient() {
        return this._telemetryClient;
    }
}
exports.ComponentDialog = ComponentDialog;
