// botframework-connector/lib/connectorApi/models/parameters.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.activityId = {
    parameterPath: "activityId",
    mapper: {
        required: true,
        serializedName: "activityId",
        type: {
            name: "String"
        }
    }
};
exports.attachmentId = {
    parameterPath: "attachmentId",
    mapper: {
        required: true,
        serializedName: "attachmentId",
        type: {
            name: "String"
        }
    }
};
exports.continuationToken = {
    parameterPath: [
        "options",
        "continuationToken"
    ],
    mapper: {
        serializedName: "continuationToken",
        type: {
            name: "String"
        }
    }
};
exports.conversationId = {
    parameterPath: "conversationId",
    mapper: {
        required: true,
        serializedName: "conversationId",
        type: {
            name: "String"
        }
    }
};
exports.memberId = {
    parameterPath: "memberId",
    mapper: {
        required: true,
        serializedName: "memberId",
        type: {
            name: "String"
        }
    }
};
exports.pageSize = {
    parameterPath: [
        "options",
        "pageSize"
    ],
    mapper: {
        serializedName: "pageSize",
        type: {
            name: "Number"
        }
    }
};
exports.viewId = {
    parameterPath: "viewId",
    mapper: {
        required: true,
        serializedName: "viewId",
        type: {
            name: "String"
        }
    }
};
