// botbuilder-dialogs/lib/choices/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @private
 */
exports.channels = {
    console: 'console',
    cortana: 'cortana',
    directline: 'directline',
    email: 'email',
    emulator: 'emulator',
    facebook: 'facebook',
    groupme: 'groupme',
    kik: 'kik',
    line: 'line',
    msteams: 'msteams',
    skype: 'skype',
    skypeforbusiness: 'skypeforbusiness',
    slack: 'slack',
    sms: 'sms',
    telegram: 'telegram',
    webchat: 'webchat'
};
/**
 * @private
 * @param channelId the id of a channel
 * @param buttonCnt count of buttons allowed
 */
function supportsSuggestedActions(channelId, buttonCnt = 100) {
    switch (channelId) {
        case exports.channels.facebook:
        case exports.channels.skype:
            return (buttonCnt <= 10);
        case exports.channels.line:
            return (buttonCnt <= 13);
        case exports.channels.kik:
            return (buttonCnt <= 20);
        case exports.channels.slack:
        case exports.channels.telegram:
        case exports.channels.emulator:
        case exports.channels.directline:
        case exports.channels.webchat:
            return (buttonCnt <= 100);
        default:
            return false;
    }
}
exports.supportsSuggestedActions = supportsSuggestedActions;
/**
 * @private
 * @param channelId the id of a channel
 * @param buttonCnt count of buttons allowed
 */
function supportsCardActions(channelId, buttonCnt = 100) {
    switch (channelId) {
        case exports.channels.facebook:
        case exports.channels.skype:
        case exports.channels.msteams:
            return (buttonCnt <= 3);
        case exports.channels.line:
            return (buttonCnt <= 99);
        case exports.channels.slack:
        case exports.channels.emulator:
        case exports.channels.directline:
        case exports.channels.webchat:
        case exports.channels.cortana:
            return (buttonCnt <= 100);
        default:
            return false;
    }
}
exports.supportsCardActions = supportsCardActions;
/**
 * @private
 * @param channelId id of a channel
 */
function hasMessageFeed(channelId) {
    switch (channelId) {
        case exports.channels.cortana:
            return false;
        default:
            return true;
    }
}
exports.hasMessageFeed = hasMessageFeed;
/**
 * @private
 * @param channelId id of a channel
 */
function maxActionTitleLength(channelId) {
    return 20;
}
exports.maxActionTitleLength = maxActionTitleLength;
/**
 * @private
 * @param context a TurnContext object representing an incoming message
 */
function getChannelId(context) {
    return context.activity.channelId || '';
}
exports.getChannelId = getChannelId;
