// botframework-streaming/lib/payloads/payloadAssemblerManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const payloadAssembler_1 = require("../assemblers/payloadAssembler");
const payloadTypes_1 = require("./payloadTypes");
/**
 * Orchestrates assemly of payloads
 */
class PayloadAssemblerManager {
    constructor(streamManager, onReceiveResponse, onReceiveRequest) {
        this.activeAssemblers = {};
        this.streamManager = streamManager;
        this.onReceiveRequest = onReceiveRequest;
        this.onReceiveResponse = onReceiveResponse;
    }
    getPayloadStream(header) {
        if (header.payloadType === payloadTypes_1.PayloadTypes.stream) {
            return this.streamManager.getPayloadStream(header);
        }
        else if (!this.activeAssemblers[header.id]) {
            let assembler = this.createPayloadAssembler(header);
            if (assembler) {
                this.activeAssemblers[header.id] = assembler;
                return assembler.getPayloadStream();
            }
        }
    }
    onReceive(header, contentStream, contentLength) {
        if (header.payloadType === payloadTypes_1.PayloadTypes.stream) {
            this.streamManager.onReceive(header, contentStream, contentLength);
        }
        else {
            if (this.activeAssemblers && this.activeAssemblers[header.id]) {
                let assembler = this.activeAssemblers[header.id];
                assembler.onReceive(header, contentStream, contentLength);
            }
            if (header.end) {
                delete this.activeAssemblers[header.id];
            }
        }
    }
    createPayloadAssembler(header) {
        if (header.payloadType === payloadTypes_1.PayloadTypes.request) {
            return new payloadAssembler_1.PayloadAssembler(this.streamManager, { header: header, onCompleted: this.onReceiveRequest });
        }
        else if (header.payloadType === payloadTypes_1.PayloadTypes.response) {
            return new payloadAssembler_1.PayloadAssembler(this.streamManager, { header: header, onCompleted: this.onReceiveResponse });
        }
    }
}
exports.PayloadAssemblerManager = PayloadAssemblerManager;
