// botframework-connector/lib/tokenApi/models/mappers.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenResponse = {
    serializedName: "TokenResponse",
    type: {
        name: "Composite",
        className: "TokenResponse",
        modelProperties: {
            channelId: {
                serializedName: "channelId",
                type: {
                    name: "String"
                }
            },
            connectionName: {
                serializedName: "connectionName",
                type: {
                    name: "String"
                }
            },
            token: {
                serializedName: "token",
                type: {
                    name: "String"
                }
            },
            expiration: {
                serializedName: "expiration",
                type: {
                    name: "String"
                }
            }
        }
    }
};
exports.InnerHttpError = {
    serializedName: "InnerHttpError",
    type: {
        name: "Composite",
        className: "InnerHttpError",
        modelProperties: {
            statusCode: {
                serializedName: "statusCode",
                type: {
                    name: "Number"
                }
            },
            body: {
                serializedName: "body",
                type: {
                    name: "Object"
                }
            }
        }
    }
};
exports.ErrorModel = {
    serializedName: "Error",
    type: {
        name: "Composite",
        className: "ErrorModel",
        modelProperties: {
            code: {
                serializedName: "code",
                type: {
                    name: "String"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            },
            innerHttpError: {
                serializedName: "innerHttpError",
                type: {
                    name: "Composite",
                    className: "InnerHttpError"
                }
            }
        }
    }
};
exports.ErrorResponse = {
    serializedName: "ErrorResponse",
    type: {
        name: "Composite",
        className: "ErrorResponse",
        modelProperties: {
            error: {
                serializedName: "error",
                type: {
                    name: "Composite",
                    className: "ErrorModel"
                }
            }
        }
    }
};
exports.AadResourceUrls = {
    serializedName: "AadResourceUrls",
    type: {
        name: "Composite",
        className: "AadResourceUrls",
        modelProperties: {
            resourceUrls: {
                serializedName: "resourceUrls",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
exports.TokenStatus = {
    serializedName: "TokenStatus",
    type: {
        name: "Composite",
        className: "TokenStatus",
        modelProperties: {
            channelId: {
                serializedName: "channelId",
                type: {
                    name: "String"
                }
            },
            connectionName: {
                serializedName: "connectionName",
                type: {
                    name: "String"
                }
            },
            hasToken: {
                serializedName: "hasToken",
                type: {
                    name: "Boolean"
                }
            },
            serviceProviderDisplayName: {
                serializedName: "serviceProviderDisplayName",
                type: {
                    name: "String"
                }
            }
        }
    }
};
