// botframework-connector/lib/tokenApi/models/parameters.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.channelId = {
    parameterPath: [
        "options",
        "channelId"
    ],
    mapper: {
        serializedName: "channelId",
        type: {
            name: "String"
        }
    }
};
exports.code = {
    parameterPath: [
        "options",
        "code"
    ],
    mapper: {
        serializedName: "code",
        type: {
            name: "String"
        }
    }
};
exports.codeChallenge = {
    parameterPath: [
        "options",
        "codeChallenge"
    ],
    mapper: {
        serializedName: "code_challenge",
        type: {
            name: "String"
        }
    }
};
exports.connectionName0 = {
    parameterPath: "connectionName",
    mapper: {
        required: true,
        serializedName: "connectionName",
        type: {
            name: "String"
        }
    }
};
exports.connectionName1 = {
    parameterPath: [
        "options",
        "connectionName"
    ],
    mapper: {
        serializedName: "connectionName",
        type: {
            name: "String"
        }
    }
};
exports.emulatorUrl = {
    parameterPath: [
        "options",
        "emulatorUrl"
    ],
    mapper: {
        serializedName: "emulatorUrl",
        type: {
            name: "String"
        }
    }
};
exports.finalRedirect = {
    parameterPath: [
        "options",
        "finalRedirect"
    ],
    mapper: {
        serializedName: "finalRedirect",
        type: {
            name: "String"
        }
    }
};
exports.include = {
    parameterPath: [
        "options",
        "include"
    ],
    mapper: {
        serializedName: "include",
        type: {
            name: "String"
        }
    }
};
exports.state = {
    parameterPath: "state",
    mapper: {
        required: true,
        serializedName: "state",
        type: {
            name: "String"
        }
    }
};
exports.userId = {
    parameterPath: "userId",
    mapper: {
        required: true,
        serializedName: "userId",
        type: {
            name: "String"
        }
    }
};
