// botframework-streaming/lib/payloadTransport/payloadSender.js
"use strict";
/**
 * @module botframework-streaming
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const headerSerializer_1 = require("../payloads/headerSerializer");
const payloadConstants_1 = require("../payloads/payloadConstants");
const transportDisconnectedEvent_1 = require("./transportDisconnectedEvent");
/**
 * Streaming payload sender.
 */
class PayloadSender {
    /**
     * Tests whether the transport sender is connected.
     *
     * @returns true if connected to a transport sender.
     */
    get isConnected() {
        return !!this.sender;
    }
    /**
     * Connects to the given transport sender.
     *
     * @param sender The transport sender to connect this payload sender to.
     */
    connect(sender) {
        this.sender = sender;
    }
    /**
     * Sends a payload out over the connected transport sender.
     *
     * @param header The header to attach to the outgoing payload.
     * @param payload The stream of buffered data to send.
     * @param sentCalback The function to execute when the send has completed.
     */
    sendPayload(header, payload, sentCallback) {
        var packet = { header, payload, sentCallback };
        this.writePacket(packet);
    }
    /**
     * Disconnects this payload sender.
     *
     * @param e The disconnected event arguments to include in the disconnected event broadcast.
     */
    disconnect(e) {
        if (this.isConnected) {
            this.sender.close();
            this.sender = null;
            if (this.disconnected) {
                this.disconnected(this, e || transportDisconnectedEvent_1.TransportDisconnectedEvent.Empty);
            }
        }
    }
    writePacket(packet) {
        try {
            let sendHeaderBuffer = Buffer.alloc(payloadConstants_1.PayloadConstants.MaxHeaderLength);
            headerSerializer_1.HeaderSerializer.serialize(packet.header, sendHeaderBuffer);
            this.sender.send(sendHeaderBuffer);
            if (packet.header.payloadLength > 0 && packet.payload) {
                let count = packet.header.payloadLength;
                while (count > 0) {
                    let chunk = packet.payload.read(count);
                    this.sender.send(chunk);
                    count -= chunk.length;
                }
                if (packet.sentCallback) {
                    packet.sentCallback();
                }
            }
        }
        catch (e) {
            this.disconnect(new transportDisconnectedEvent_1.TransportDisconnectedEvent(e.message));
        }
    }
}
exports.PayloadSender = PayloadSender;
