// botbuilder-dialogs/lib/memory/scopes/memoryScope.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Abstract base class for all memory scopes.
 */
class MemoryScope {
    constructor(name, includeInSnapshot = true) {
        this.includeInSnapshot = includeInSnapshot;
        this.name = name;
    }
    /**
     * Changes the backing object for the memory scope.
     * @param dc Current dialog context
     * @param memory memory to assign
     */
    setMemory(dc, memory) {
        throw new Error(`MemoryScope.setMemory: The '${this.name}' memory scope is read-only.`);
    }
    /**
     * Loads a scopes backing memory at the start of a turn.
     * @param dc Current dialog context.
     */
    load(dc) {
        return __awaiter(this, void 0, void 0, function* () {
            // No initialization by default.
        });
    }
    /**
     * Saves a scopes backing memory at the end of a turn.
     * @param dc Current dialog context.
     */
    saveChanges(dc) {
        return __awaiter(this, void 0, void 0, function* () {
            // No initialization by default.
        });
    }
    /**
     * Deletes the backing memory for a scope.
     * @param dc Current dialog context.
     */
    delete(dc) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error(`MemoryScope.delete: The '${this.name}' memory scope can't be deleted.`);
        });
    }
}
exports.MemoryScope = MemoryScope;
