// botframework-connector/lib/tokenApi/tokenApiClient.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const Models = require("./models");
exports.TokenApiModels = Models;
const Mappers = require("./models/mappers");
exports.TokenApiMappers = Mappers;
const operations = require("./operations");
const tokenApiClientContext_1 = require("./tokenApiClientContext");
exports.TokenApiClientContext = tokenApiClientContext_1.TokenApiClientContext;
class TokenApiClient extends tokenApiClientContext_1.TokenApiClientContext {
    /**
     * Initializes a new instance of the TokenApiClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param [options] The parameter options
     */
    constructor(credentials, options) {
        super(credentials, options);
        this.botSignIn = new operations.BotSignIn(this);
        this.userToken = new operations.UserToken(this);
    }
}
exports.TokenApiClient = TokenApiClient;
__export(require("./operations"));
