// botframework-connector/lib/connectorApi/connectorClient.js
"use strict";
/*
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const connectorClientContext_1 = require("./connectorClientContext");
exports.ConnectorClientContext = connectorClientContext_1.ConnectorClientContext;
const Models = require("./models");
exports.ConnectorModels = Models;
const Mappers = require("./models/mappers");
exports.ConnectorMappers = Mappers;
const operations = require("./operations");
class ConnectorClient extends connectorClientContext_1.ConnectorClientContext {
    /**
     * Initializes a new instance of the ConnectorClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param [options] The parameter options
     */
    constructor(credentials, options) {
        super(credentials, options);
        this.attachments = new operations.Attachments(this);
        this.conversations = new operations.Conversations(this);
    }
}
exports.ConnectorClient = ConnectorClient;
// Operation Specifications
__export(require("./operations"));
